### DISCUSSION TITLE
Annotations in the practical use of Panopticon

### DISCUSSION THREAD

**Discussion by morphinestyle** [akeeba/panopticon#520] [Ideas]
For monitoring Panopticon is superlative, I note down some ideas and considerations made during practical use:

- **it would make a huge difference to have an "UPDATE ENTIRE SITE" button** that can completely update both the extensions and the Joomla core **FOR ONLY ONE SITE AT A TIME**, **obviously with the possibility of making a backup, both before and after the automatic operation**; in this way it is possible to update each site one at a time in complete safety with backups. At the moment I don't prefer to have the sites automatically updated (read the second point of note below), or in any case perhaps I could in the future for smaller sites, so currently Panopticon is an excellent software that gives me the possibility of finally having good monitoring of all the sites, but in practice I have not found a way to launch a "full" update of a site in an automated way (manual or scheduled).

- **create update templates**, because manually modifying all the extensions site by site becomes really complex; let me explain better with a practical example: minor sites could be updated in totally automatic mode, while important sites should be updated automatically at most only with patch updates, currently these customizations should be done manually site-by-site extension-per-extension and it would be really complex... If update templates existed, you could create update schemes (patch, minor, version) and easily assign them to a website based on the importance of the website itself: an "automatic-update-all-template" to less important website and an “update-patch-only template” to the more important website

- **possibility of specifying the email addresses to which to send the Automated Emails**: Scheduled Update Summary and Scheduled Action Summary (in this field I could put the email of my customers, **who are not interested in receiving all the other technical emails**, but only one summary every month)

- in the site overview: be **able to sort based on the joomla version** (so that the site that needs to be updated first is immediately visible)

- in the overview of sites: **be able to sort based on the number of extensions needs to be update** (so as to be immediately visible which site needs to be updated first)

- in the overview screen, **align the names of the sites regardless of the presence and size of the favicons**: currently they are all messy and make identification slower because they worsen visibility

- possibility to remove the update notice for PHP patch versions: for example using lsphp 8.2.11 for each site asks everyone to update to 8.2.14

- an overview page showing the list of expiry dates for the sites' SSL/TLS certificates

**Comment by nikosdion** [comment]
Please avoid huge discussions with a dozen topics. They are impossible to track. At least, I cannot track them, which is why I am only going to reply once here.

Update entire site: Such a button would be unclear on what it will do. Does it update Joomla, or extensions? Does it respect automatic update version limits, or not? If it does, what is the point? If it does not, what is the difference to using the Core and Extension Updates overview pages? It's an internally inconsistent feature. At best I could add a “YOLO!” button in the extensions list, updating _everything_ regardless of your automatic update preferences. With a big, red, all caps warning that what you are about to do is very likely to make your site go ka-blewey. I have been tempted to do that. You know what? I can do that, but put it behind a Global Configuration and site-specific option — just like I have done with other dangerous options in other software, e.g. the option to delete all existing files and folders before integrated restoration in Akeeba Backup (if you didn't know the option existed, congrats, you prove my point; you have to read the docs to know how to enable a dangerous option!).

Update template: Hard no.

Here is the fundamental concept of how to use Panopticon. You set up each site _once_ to automatically update core and extensions using the limits you deem appropriate for each site. What is left is by definition something (core or extension) you **MUST** test manually if it can be updated before updating, e.g. make a clone with Akeeba Backup, apply the update, and figure out if it updates cleanly. In this case, and assuming it updates cleanly, you can apply the updates manually a. using the Site Information page of one site, b. the Core and Extension Updates Overview pages, or c. the backend of each site. For the first two cases, assuming you have _similar_ sites you are okay mass-updating after having only tested the updates on one of them, you can use **groups**. Groups without any permissions are essentially _tags_. You can filter by these groups in the overview pages. You can also filter by Joomla version, and extension information. Therefore, you can mass-apply updates of Joomla! or specific extensions to all sites in a group with a few clicks: select which sites / extensions you want to update, click on the toolbar button to schedule the updates.

If you need something more complex you have crossed the threshold beyond which using a GUI is more time intensive than setting up a proper devops solution with Ansible (or Chef, or Puppet, or…) and the Joomla! CLI application. I am not interested in reproducing the mistakes Microsoft and Apple did in the 90s and early 2010s.

Email addresses: I will add a field in the scheduled task configuration to enter the email addresses you want to send the emails. This requires some refactoring of the email sending which is why I did not do that in 1.1.0.

Sort by Joomla version: sorting by a value stored in a JSON field is _slow_ unless I could raise the requirements to MySQL 8.0 or later and MariaDB 11.0 or later, thereby excluding the majority of users. Moreover, that is an alpha sort which means that for Joomla 3 the sorted ascending list is 3.1, 3.10, 3.2, etc. Sure, with Joomla! minor versions only reaching to 4 it's not a problem… unless you take into account the patch versions. Once we hit a x.y.10 version it will be sorted between x.y.1 and x.y.2. Using version sorting requires loading all records in memory which means that you run out of PHP memory or need several seconds (instead of a few milliseconds) to display each page. So, that's a hard nope. The solution to that is the Core Updates overview page.

Sort by number of extensions with updates: You cannot sort by a computed field unless you load all records at once, in which case you will run out of PHP memory and make page load take several seconds instead of a 20 milliseconds. So, that's a hard nope. The solution to that is the Extension Updates overview page.

Remove the update notice for PHP patch versions: You have missed the entire point. You are using an out of date version of PHP **with known bugs and security issues**. So, yes, **you do need to update**. If PHP patch version updates were unimportant _they would not be released_. There would be a new PHP version once every two years and nothing in between.

Overview page showing the list of expiry dates for the sites' SSL/TLS certificates: Please open an issue and explain your use case.

**Comment by morphinestyle** [comment]
Ok, you were clear.

At least consider adding a button to make it easy to bulk customize your update preference to long lists of extensions - some sites have over 70 extensions that you have to manually edit!

However, a question arises: if globally I set to update all the extensions, but for a site I set not to do automatic updates to the extensions: what happens if I add a new extension to that website? How will the new extension found be set by default?

Regarding alignment and favicons, here is an example:
![image](https://github.com/akeeba/panopticon/assets/35969533/18e65e36-0f1d-4671-9437-ae6f3e387134)

**Comment by nikosdion** [comment]
> At least consider adding a button to make it easy to bulk customize your update preference to long lists of extensions - some sites have over 70 extensions that you have to manually edit!

There's not much we can do there. It comes down to figuring out your default preferences and choosing settings for each site.

> However, a question arises: if globally I set to update all the extensions, but for a site I set not to do automatic updates to the extensions: what happens if I add a new extension to that website? How will the new extension found be set by default?

Remember how the preferences cascade? Whatever is explicitly set for the extension on the site. If it's set to use global (default) it takes the preference set in System Configuration for this extension. If that is set to use global it falls back to the global "When An Extension Update Is Found" preference which is by default set to None.

So, to answer your question: it depends on your global settings.

This should inform your strategy. I prefer things to _fail safe_. My global "When An Extension Update Is Found" preference is Send Email because it's the absolute safest thing to do. It won't sit there ignoring the extension, nor will it install a version I did not expect. It will just send me an email. I will receive the email and go "d'oh, I need to configure this".

When there are _safe_ extensions, e.g. Akeeba Backup and Admin Tools and JCE, whose updates from patch to major are 100% safe I go to the system configuration page and set them to update automatically to major versions.

For everything else, I evaluate the extension in the context of the site. Is it safe to auto-update it, and if so between which versions? There's a calendar extension which changes template layouts from one minor version to the other. I auto-update patch versions only.

> Regarding alignment and favicons, here is an example:

Yes, I know. I am going to add a fixed width to the container.
